packages/react-i18n/src/babel-plugin.ts

```diff
@@ -0,0 +1,155 @@
+import {TemplateBuilder} from '@babel/template';
+import * as Types from '@babel/types';
+import {NodePath} from '@babel/traverse';
+import stringHash from 'string-hash';
+import path from 'path';
+import fs from 'fs';
+
+interface State {
+  lastImport: NodePath<Types.ImportDeclaration>;
+}
+
+export default function injectWithI18nArguments({
+  types: t,
+  template,
+}: {
+  types: typeof Types;
+  template: TemplateBuilder<Types.ImportDeclaration | Types.ObjectExpression>;
+}) {
+  function fallbackTranslationsImport({id}) {
+    return template(`import ${id} from './translations/en.json';`, {
+      sourceType: 'module',
+    })();
+  }
+
+  function i18nCallExpression({id, fallbackID, decoratorName}) {
+    return template(
+      `${decoratorName}({
+      id: '${id}',
+      fallback: ${fallbackID},
+      async translations(locale) {
+        return await import(/* webpackChunkName: '${id}-i18n' */ \`./translations/$\{locale}.json\`);
+      },
+    })`,
+      {
+        sourceType: 'module',
+        plugins: ['dynamicImport'],
+        preserveComments: true,
+      },
+    )();
+  }
+
+  function addI18nArguments({
+    binding,
+    bindingName,
+    filename,
+    filenameRelative,
+    fallbackID,
+    insertImport,
+  }) {
+    let _hasTranslations;
+
+    for (const refPath of binding.referencePaths) {
+      const callExpression: NodePath = refPath.parentPath;
+
+      if (!callExpression.isCallExpression()) {
+        continue;
+      }
+
+      const args = callExpression.get('arguments');
+      if (args.length !== 0) {
+        continue;
+      }
+
+      if (_hasTranslations == null) {
+        _hasTranslations = hasTranslations(filename);
+        if (_hasTranslations) {
+          insertImport();
+        }
+      }
+
+      if (!_hasTranslations) {
+        return;
+      }
+
+      callExpression.replaceWith(
+        i18nCallExpression({
+          id: generateID(filenameRelative),
+          fallbackID,
+          decoratorName: bindingName,
+        }),
+      );
+    }
+  }
+
+  return {
+    visitor: {
+      Program(nodePath: NodePath<Types.Program>, state: State) {
+        const lastImport = nodePath
+          .get('body')
+          .filter(subPath => subPath.isImportDeclaration())
+          .pop();
+        state.lastImport = lastImport as NodePath<Types.ImportDeclaration>;
+      },
+      ImportDeclaration(
+        nodePath: NodePath<Types.ImportDeclaration>,
+        state: State,
+      ) {
+        if (nodePath.node.source.value !== '@shopify/react-i18n') {
+          return;
+        }
+        const {specifiers} = nodePath.node;
+        for (const specifier of specifiers) {
+          if (
```

Comment:
Might be better to flip the condition and `continue` if false? There is a lot of nesting that goes on here